File: KKY_Tool_Revit_2019-2023/Infrastructure/ExcelExportStyleRegistry.vb
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.IO
Imports NPOI.SS.UserModel
Imports NPOI.SS.Util
Imports NPOI.XSSF.UserModel

Namespace Infrastructure

    Public Module ExcelExportStyleRegistry

        Public Delegate Function RowStatusResolver(row As DataRow, table As DataTable) As ExcelStyleHelper.RowStatus

        Private ReadOnly _resolvers As New Dictionary(Of String, RowStatusResolver)(StringComparer.OrdinalIgnoreCase)

        Sub New()
            Register("connector", AddressOf ResolveConnector)
            Register("guid", AddressOf ResolveResultLike)
            Register("paramprop", AddressOf ResolveParamProp)
            Register("points", AddressOf ResolveResultLike)
            Register("pms", AddressOf ResolveResultLike)
            Register("familylink", AddressOf ResolveIssueLike)
            Register("sharedparambatch", AddressOf ResolveSharedParamBatch)

            Register("connector diagnostics", AddressOf ResolveConnector)
            Register("familylinkaudit", AddressOf ResolveIssueLike)
            Register("family link audit", AddressOf ResolveIssueLike)
            Register("pms vs segment size검토", AddressOf ResolveResultLike)
            Register("pipe segment class검토", AddressOf ResolveResultLike)
            Register("routing class검토", AddressOf ResolveResultLike)
        End Sub

        Public Sub Register(key As String, resolver As RowStatusResolver)
            If String.IsNullOrWhiteSpace(key) OrElse resolver Is Nothing Then Return
            _resolvers(key.Trim()) = resolver
        End Sub

        Public Function Resolve(sheetNameOrKey As String, row As DataRow, table As DataTable) As ExcelStyleHelper.RowStatus
            If row Is Nothing OrElse table Is Nothing Then Return ExcelStyleHelper.RowStatus.None

            Dim key = NormalizeKey(sheetNameOrKey)
            Dim fn As RowStatusResolver = Nothing

            If Not String.IsNullOrWhiteSpace(key) AndAlso _resolvers.TryGetValue(key, fn) Then
                Return SafeResolve(fn, row, table)
            End If

            If Not String.IsNullOrWhiteSpace(sheetNameOrKey) AndAlso _resolvers.TryGetValue(sheetNameOrKey.Trim(), fn) Then
                Return SafeResolve(fn, row, table)
            End If

            Return ResolveGeneric(row, table)
        End Function

        Public Function FilterIssueRows(styleKey As String, table As DataTable) As DataTable
            If table Is Nothing Then Return Nothing

            Dim filtered As DataTable = table.Clone()
            For Each row As DataRow In table.Rows
                Dim status As ExcelStyleHelper.RowStatus = Resolve(styleKey, row, table)
                If status = ExcelStyleHelper.RowStatus.Warning OrElse status = ExcelStyleHelper.RowStatus.[Error] Then
                    filtered.ImportRow(row)
                End If
            Next
            Return filtered
        End Function

        Public Sub ApplyStylesForKey(styleKey As String,
                                     xlsxPath As String,
                                     Optional sheetName As String = Nothing,
                                     Optional autoFit As Boolean = True,
                                     Optional excelMode As String = "normal")
            If String.IsNullOrWhiteSpace(xlsxPath) Then Return
            If Not File.Exists(xlsxPath) Then Return

            Using fs As New FileStream(xlsxPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite)
                Dim wb As IWorkbook = New XSSFWorkbook(fs)

                Dim targetSheets As New List(Of ISheet)()
                If Not String.IsNullOrWhiteSpace(sheetName) Then
                    Dim namedSheet = wb.GetSheet(sheetName)
                    If namedSheet IsNot Nothing Then targetSheets.Add(namedSheet)
                End If

                If targetSheets.Count = 0 Then
                    For i As Integer = 0 To wb.NumberOfSheets - 1
                        Dim sh = wb.GetSheetAt(i)
                        If sh IsNot Nothing Then targetSheets.Add(sh)
                    Next
                End If

                For Each sh In targetSheets
                    ApplyStandardSheetStyle(wb, sh)
                    ApplyBorders(wb, sh)
                    ApplyKeySpecificFormats(styleKey, sh)
                Next

                fs.Close()
                Using outFs As New FileStream(xlsxPath, FileMode.Create, FileAccess.Write, FileShare.None)
                    wb.Write(outFs)
                End Using
            End Using
        End Sub

        Private Sub ApplyStandardSheetStyle(wb As IWorkbook, sh As ISheet)
            If wb Is Nothing OrElse sh Is Nothing Then Return

            Dim header = sh.GetRow(0)
            If header Is Nothing Then Return

            Dim headerStyle As ICellStyle = ExcelStyleHelper.GetHeaderStyle(wb)
            If headerStyle Is Nothing Then Return

            Dim lastCol As Integer = CInt(header.LastCellNum) - 1
            If lastCol < 0 Then Return

            For c As Integer = 0 To lastCol
                Dim cell = header.GetCell(c)
                If cell Is Nothing Then
                    cell = header.CreateCell(c)
                    cell.SetCellValue("")
                End If
                cell.CellStyle = headerStyle
            Next
        End Sub

        Private Sub ApplyBorders(wb As IWorkbook, sh As ISheet)
            If wb Is Nothing OrElse sh Is Nothing Then Return
            Dim lastRow As Integer = sh.LastRowNum
            If lastRow < 0 Then Return

            For r As Integer = 0 To lastRow
                Dim row = sh.GetRow(r)
                If row Is Nothing Then Continue For

                Dim lastCol As Integer = CInt(row.LastCellNum) - 1
                If lastCol < 0 Then Continue For

                For c As Integer = 0 To lastCol
                    Dim cell = row.GetCell(c)
                    If cell Is Nothing Then
                        cell = row.CreateCell(c)
                        cell.SetCellValue("")
                    End If

                    Dim src = cell.CellStyle
                    Dim dst = wb.CreateCellStyle()
                    If src IsNot Nothing Then
                        dst.CloneStyleFrom(src)
                    End If
                    dst.BorderBottom = BorderStyle.Thin
                    dst.BorderTop = BorderStyle.Thin
                    dst.BorderLeft = BorderStyle.Thin
                    dst.BorderRight = BorderStyle.Thin
                    cell.CellStyle = dst
                Next
            Next
        End Sub

        Private Sub ApplyKeySpecificFormats(styleKey As String, sh As ISheet)
            If sh Is Nothing Then Return
            Dim key As String = NormalizeKey(styleKey)
            If String.IsNullOrWhiteSpace(key) Then Return

            ' key별 세부 포맷 확장 포인트 (현재는 중복 처리 방지를 위해 최소 동작 유지)
            If key = "pms" Then
                Return
            End If
        End Sub

        Private Function SafeResolve(fn As RowStatusResolver, row As DataRow, table As DataTable) As ExcelStyleHelper.RowStatus
            Try
                Return fn(row, table)
            Catch
                Return ExcelStyleHelper.RowStatus.None
            End Try
        End Function

        Private Function NormalizeKey(sheetNameOrKey As String) As String
            If String.IsNullOrWhiteSpace(sheetNameOrKey) Then Return ""
            Dim s = sheetNameOrKey.Trim().ToLowerInvariant()
            If s.Contains("connector") Then Return "connector"
            If s.Contains("guid") Then Return "guid"
            If s.Contains("param") Then Return "paramprop"
            If s.Contains("point") Then Return "points"
            If s.Contains("sharedparambatch") Then Return "sharedparambatch"
            If s.Contains("familylink") OrElse s.Contains("family link") Then Return "familylink"
            If s.Contains("pms") OrElse s.Contains("segment") Then Return "pms"
            Return s
        End Function

        Private Function ResolveConnector(row As DataRow, table As DataTable) As ExcelStyleHelper.RowStatus
            Dim statusText = GetFirstExistingText(row, table, "Status", "Result", "검토결과")
            If IsOkLike(statusText) Then Return ExcelStyleHelper.RowStatus.None
            If LooksError(statusText) Then Return ExcelStyleHelper.RowStatus.[Error]
            Return ExcelStyleHelper.RowStatus.Warning
        End Function

        Private Function ResolveIssueLike(row As DataRow, table As DataTable) As ExcelStyleHelper.RowStatus
            Dim issue = GetFirstExistingText(row, table, "Issue", "Result", "Status")
            If IsOkLike(issue) Then Return ExcelStyleHelper.RowStatus.None
            If LooksError(issue) Then Return ExcelStyleHelper.RowStatus.[Error]
            Return ExcelStyleHelper.RowStatus.Warning
        End Function

        Private Function ResolveResultLike(row As DataRow, table As DataTable) As ExcelStyleHelper.RowStatus
            Dim result = GetFirstExistingText(row, table, "Result", "Res", "Status", "성공여부", "검토결과", "Class검토결과", "Class검토")
            If IsOkLike(result) Then Return ExcelStyleHelper.RowStatus.None
            If LooksError(result) Then Return ExcelStyleHelper.RowStatus.[Error]
            If String.IsNullOrWhiteSpace(result) Then Return ExcelStyleHelper.RowStatus.Info
            Return ExcelStyleHelper.RowStatus.Warning
        End Function

        Private Function ResolveParamProp(row As DataRow, table As DataTable) As ExcelStyleHelper.RowStatus
            Dim detail = GetFirstExistingText(row, table, "Detail", "Result", "Status", "메시지")
            If String.IsNullOrWhiteSpace(detail) Then Return ExcelStyleHelper.RowStatus.None
            If IsOkLike(detail) Then Return ExcelStyleHelper.RowStatus.None
            If LooksError(detail) Then Return ExcelStyleHelper.RowStatus.[Error]
            Return ExcelStyleHelper.RowStatus.Warning
        End Function

        Private Function ResolveSharedParamBatch(row As DataRow, table As DataTable) As ExcelStyleHelper.RowStatus
            Dim status = GetFirstExistingText(row, table, "성공여부", "Level", "Status", "Result", "메시지")
            If IsOkLike(status) Then Return ExcelStyleHelper.RowStatus.None
            If LooksError(status) Then Return ExcelStyleHelper.RowStatus.[Error]
            If String.IsNullOrWhiteSpace(status) Then Return ExcelStyleHelper.RowStatus.None
            Return ExcelStyleHelper.RowStatus.Warning
        End Function

        Private Function ResolveGeneric(row As DataRow, table As DataTable) As ExcelStyleHelper.RowStatus
            Dim txt = GetFirstExistingText(row, table, "Status", "Result", "Issue", "Error", "ErrorMessage", "Notes")
            If String.IsNullOrWhiteSpace(txt) Then Return ExcelStyleHelper.RowStatus.None
            If IsOkLike(txt) Then Return ExcelStyleHelper.RowStatus.None
            If LooksError(txt) Then Return ExcelStyleHelper.RowStatus.[Error]
            Return ExcelStyleHelper.RowStatus.Warning
        End Function

        Private Function GetFirstExistingText(row As DataRow, table As DataTable, ParamArray colNames() As String) As String
            For Each c In colNames
                Dim t = GetColText(row, table, c)
                If Not String.IsNullOrWhiteSpace(t) Then Return t
            Next
            Return ""
        End Function

        Private Function GetColText(row As DataRow, table As DataTable, colName As String) As String
            For Each col As DataColumn In table.Columns
                If String.Equals(col.ColumnName, colName, StringComparison.OrdinalIgnoreCase) Then
                    Dim v = row(col)
                    If v Is Nothing OrElse v Is DBNull.Value Then Return ""
                    Return v.ToString().Trim()
                End If
            Next
            Return ""
        End Function

        Private Function IsOkLike(s As String) As Boolean
            If String.IsNullOrWhiteSpace(s) Then Return False
            Dim t = s.Trim().ToLowerInvariant()
            If t = "ok" OrElse t = "pass" OrElse t = "success" Then Return True
            If t.StartsWith("ok(") OrElse t.StartsWith("ok[") OrElse t.StartsWith("ok_") Then Return True
            If t.Contains("오류 없음") OrElse t.Contains("정상") OrElse t.Contains("이상 없음") Then Return True
            Return False
        End Function

        Private Function LooksError(s As String) As Boolean
            If String.IsNullOrWhiteSpace(s) Then Return False
            Dim t = s.Trim().ToLowerInvariant()
            If t.Contains("mismatch") OrElse t.Contains("불일치") Then Return True
            If t.Contains("error") OrElse t.Contains("fail") Then Return True
            If t.Contains("실패") OrElse t.Contains("오류") Then Return True
            Return False
        End Function

    End Module

End Namespace

File: KKY_Tool_Revit_2019-2023/Infrastructure/ExcelExportStyleRegistry.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.IO;
using NPOI.SS.UserModel;
using NPOI.XSSF.UserModel;

namespace KKY_Tool_Revit.Infrastructure
{
    public static class ExcelExportStyleRegistry
    {
        public delegate ExcelStyleHelper.RowStatus RowStatusResolver(DataRow row, DataTable table);

        private static readonly Dictionary<string, RowStatusResolver> Resolvers = new Dictionary<string, RowStatusResolver>(StringComparer.OrdinalIgnoreCase);

        static ExcelExportStyleRegistry()
        {
            Register("connector", ResolveConnector);
            Register("guid", ResolveResultLike);
            Register("paramprop", ResolveParamProp);
            Register("points", ResolveResultLike);
            Register("pms", ResolveResultLike);
            Register("familylink", ResolveIssueLike);
            Register("sharedparambatch", ResolveSharedParamBatch);

            Register("connector diagnostics", ResolveConnector);
            Register("familylinkaudit", ResolveIssueLike);
            Register("family link audit", ResolveIssueLike);
            Register("pms vs segment size검토", ResolveResultLike);
            Register("pipe segment class검토", ResolveResultLike);
            Register("routing class검토", ResolveResultLike);
        }

        public static void Register(string key, RowStatusResolver resolver)
        {
            if (string.IsNullOrWhiteSpace(key) || resolver == null) return;
            Resolvers[key.Trim()] = resolver;
        }

        public static ExcelStyleHelper.RowStatus Resolve(string sheetNameOrKey, DataRow row, DataTable table)
        {
            if (row == null || table == null) return ExcelStyleHelper.RowStatus.None;

            var key = NormalizeKey(sheetNameOrKey);
            if (!string.IsNullOrWhiteSpace(key) && Resolvers.TryGetValue(key, out var fn))
            {
                return SafeResolve(fn, row, table);
            }

            if (!string.IsNullOrWhiteSpace(sheetNameOrKey) && Resolvers.TryGetValue(sheetNameOrKey.Trim(), out fn))
            {
                return SafeResolve(fn, row, table);
            }

            return ResolveGeneric(row, table);
        }

        public static DataTable FilterIssueRows(string styleKey, DataTable table)
        {
            if (table == null) return null;

            var filtered = table.Clone();
            foreach (DataRow row in table.Rows)
            {
                var status = Resolve(styleKey, row, table);
                if (status == ExcelStyleHelper.RowStatus.Warning || status == ExcelStyleHelper.RowStatus.Error)
                {
                    filtered.ImportRow(row);
                }
            }
            return filtered;
        }

        public static void ApplyStylesForKey(string styleKey, string xlsxPath, string sheetName = null, bool autoFit = true, string excelMode = "normal")
        {
            if (string.IsNullOrWhiteSpace(xlsxPath)) return;
            if (!File.Exists(xlsxPath)) return;

            using (var fs = new FileStream(xlsxPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
            {
                IWorkbook wb = new XSSFWorkbook(fs);

                var targetSheets = new List<ISheet>();
                if (!string.IsNullOrWhiteSpace(sheetName))
                {
                    var namedSheet = wb.GetSheet(sheetName);
                    if (namedSheet != null) targetSheets.Add(namedSheet);
                }

                if (targetSheets.Count == 0)
                {
                    for (var i = 0; i < wb.NumberOfSheets; i++)
                    {
                        var sh = wb.GetSheetAt(i);
                        if (sh != null) targetSheets.Add(sh);
                    }
                }

                foreach (var sh in targetSheets)
                {
                    ApplyStandardSheetStyle(wb, sh);
                    ApplyBorders(wb, sh);
                    ApplyKeySpecificFormats(styleKey, sh);
                }

                fs.Close();
                using (var outFs = new FileStream(xlsxPath, FileMode.Create, FileAccess.Write, FileShare.None))
                {
                    wb.Write(outFs);
                }
            }
        }

        private static void ApplyStandardSheetStyle(IWorkbook wb, ISheet sh)
        {
            if (wb == null || sh == null) return;

            var header = sh.GetRow(0);
            if (header == null) return;

            var headerStyle = ExcelStyleHelper.GetHeaderStyle(wb);
            if (headerStyle == null) return;

            var lastCol = header.LastCellNum - 1;
            if (lastCol < 0) return;

            for (var c = 0; c <= lastCol; c++)
            {
                var cell = header.GetCell(c) ?? header.CreateCell(c);
                if (cell.CellType == CellType.Blank) cell.SetCellValue(string.Empty);
                cell.CellStyle = headerStyle;
            }
        }

        private static void ApplyBorders(IWorkbook wb, ISheet sh)
        {
            if (wb == null || sh == null) return;
            var lastRow = sh.LastRowNum;
            if (lastRow < 0) return;

            for (var r = 0; r <= lastRow; r++)
            {
                var row = sh.GetRow(r);
                if (row == null) continue;

                var lastCol = row.LastCellNum - 1;
                if (lastCol < 0) continue;

                for (var c = 0; c <= lastCol; c++)
                {
                    var cell = row.GetCell(c) ?? row.CreateCell(c);
                    if (cell.CellType == CellType.Blank) cell.SetCellValue(string.Empty);

                    var src = cell.CellStyle;
                    var dst = wb.CreateCellStyle();
                    if (src != null) dst.CloneStyleFrom(src);
                    dst.BorderBottom = BorderStyle.Thin;
                    dst.BorderTop = BorderStyle.Thin;
                    dst.BorderLeft = BorderStyle.Thin;
                    dst.BorderRight = BorderStyle.Thin;
                    cell.CellStyle = dst;
                }
            }
        }

        private static void ApplyKeySpecificFormats(string styleKey, ISheet sh)
        {
            if (sh == null) return;
            var key = NormalizeKey(styleKey);
            if (string.IsNullOrWhiteSpace(key)) return;

            if (key == "pms")
            {
                return;
            }
        }

        private static ExcelStyleHelper.RowStatus SafeResolve(RowStatusResolver fn, DataRow row, DataTable table)
        {
            try
            {
                return fn(row, table);
            }
            catch
            {
                return ExcelStyleHelper.RowStatus.None;
            }
        }

        private static string NormalizeKey(string sheetNameOrKey)
        {
            if (string.IsNullOrWhiteSpace(sheetNameOrKey)) return string.Empty;
            var s = sheetNameOrKey.Trim().ToLowerInvariant();
            if (s.Contains("connector")) return "connector";
            if (s.Contains("guid")) return "guid";
            if (s.Contains("param")) return "paramprop";
            if (s.Contains("point")) return "points";
            if (s.Contains("sharedparambatch")) return "sharedparambatch";
            if (s.Contains("familylink") || s.Contains("family link")) return "familylink";
            if (s.Contains("pms") || s.Contains("segment")) return "pms";
            return s;
        }

        private static ExcelStyleHelper.RowStatus ResolveConnector(DataRow row, DataTable table)
        {
            var statusText = GetFirstExistingText(row, table, "Status", "Result", "검토결과");
            if (IsOkLike(statusText)) return ExcelStyleHelper.RowStatus.None;
            if (LooksError(statusText)) return ExcelStyleHelper.RowStatus.Error;
            return ExcelStyleHelper.RowStatus.Warning;
        }

        private static ExcelStyleHelper.RowStatus ResolveIssueLike(DataRow row, DataTable table)
        {
            var issue = GetFirstExistingText(row, table, "Issue", "Result", "Status");
            if (IsOkLike(issue)) return ExcelStyleHelper.RowStatus.None;
            if (LooksError(issue)) return ExcelStyleHelper.RowStatus.Error;
            return ExcelStyleHelper.RowStatus.Warning;
        }

        private static ExcelStyleHelper.RowStatus ResolveResultLike(DataRow row, DataTable table)
        {
            var result = GetFirstExistingText(row, table, "Result", "Res", "Status", "성공여부", "검토결과", "Class검토결과", "Class검토");
            if (IsOkLike(result)) return ExcelStyleHelper.RowStatus.None;
            if (LooksError(result)) return ExcelStyleHelper.RowStatus.Error;
            if (string.IsNullOrWhiteSpace(result)) return ExcelStyleHelper.RowStatus.Info;
            return ExcelStyleHelper.RowStatus.Warning;
        }

        private static ExcelStyleHelper.RowStatus ResolveParamProp(DataRow row, DataTable table)
        {
            var detail = GetFirstExistingText(row, table, "Detail", "Result", "Status", "메시지");
            if (string.IsNullOrWhiteSpace(detail)) return ExcelStyleHelper.RowStatus.None;
            if (IsOkLike(detail)) return ExcelStyleHelper.RowStatus.None;
            if (LooksError(detail)) return ExcelStyleHelper.RowStatus.Error;
            return ExcelStyleHelper.RowStatus.Warning;
        }

        private static ExcelStyleHelper.RowStatus ResolveSharedParamBatch(DataRow row, DataTable table)
        {
            var status = GetFirstExistingText(row, table, "성공여부", "Level", "Status", "Result", "메시지");
            if (IsOkLike(status)) return ExcelStyleHelper.RowStatus.None;
            if (LooksError(status)) return ExcelStyleHelper.RowStatus.Error;
            if (string.IsNullOrWhiteSpace(status)) return ExcelStyleHelper.RowStatus.None;
            return ExcelStyleHelper.RowStatus.Warning;
        }

        private static ExcelStyleHelper.RowStatus ResolveGeneric(DataRow row, DataTable table)
        {
            var txt = GetFirstExistingText(row, table, "Status", "Result", "Issue", "Error", "ErrorMessage", "Notes");
            if (string.IsNullOrWhiteSpace(txt)) return ExcelStyleHelper.RowStatus.None;
            if (IsOkLike(txt)) return ExcelStyleHelper.RowStatus.None;
            if (LooksError(txt)) return ExcelStyleHelper.RowStatus.Error;
            return ExcelStyleHelper.RowStatus.Warning;
        }

        private static string GetFirstExistingText(DataRow row, DataTable table, params string[] colNames)
        {
            foreach (var c in colNames)
            {
                var t = GetColText(row, table, c);
                if (!string.IsNullOrWhiteSpace(t)) return t;
            }
            return string.Empty;
        }

        private static string GetColText(DataRow row, DataTable table, string colName)
        {
            foreach (DataColumn col in table.Columns)
            {
                if (!string.Equals(col.ColumnName, colName, StringComparison.OrdinalIgnoreCase)) continue;
                var v = row[col];
                if (v == null || v is DBNull) return string.Empty;
                return v.ToString().Trim();
            }
            return string.Empty;
        }

        private static bool IsOkLike(string s)
        {
            if (string.IsNullOrWhiteSpace(s)) return false;
            var t = s.Trim().ToLowerInvariant();
            if (t == "ok" || t == "pass" || t == "success") return true;
            if (t.StartsWith("ok(") || t.StartsWith("ok[") || t.StartsWith("ok_")) return true;
            if (t.Contains("오류 없음") || t.Contains("정상") || t.Contains("이상 없음")) return true;
            return false;
        }

        private static bool LooksError(string s)
        {
            if (string.IsNullOrWhiteSpace(s)) return false;
            var t = s.Trim().ToLowerInvariant();
            if (t.Contains("mismatch") || t.Contains("불일치")) return true;
            if (t.Contains("error") || t.Contains("fail")) return true;
            if (t.Contains("실패") || t.Contains("오류")) return true;
            return false;
        }
    }
}
